mailpoet/lib/Automation/Engine/Storage/WorkflowStorage.php

```diff
@@ -26,15 +26,39 @@ public function __construct() {
   }
 
   public function createWorkflow(Workflow $workflow): int {
-    $result = $this->wpdb->insert($this->workflowTable, $this->getWorkflowHeaderData($workflow));
+    $workflowHeaderData = $this->getWorkflowHeaderData($workflow);
+    unset($workflowHeaderData['id']);
+    $result = $this->wpdb->insert($this->workflowTable, $workflowHeaderData);
     if (!$result) {
       throw Exceptions::databaseError($this->wpdb->last_error);
     }
-    $id = (int)$this->wpdb->insert_id;
+    $id = $this->wpdb->insert_id;
     $this->insertWorkflowVersion($id, $workflow);
     return $id;
   }
 
+  public function duplicateWorkflow(Workflow $workflow): int {
+    $data = $workflow->toArray();
+    $now = (new DateTimeImmutable())->format(DateTimeImmutable::W3C);
+    $data['created_at'] = $now;
+    $data['updated_at'] = $now;
+    $data['status'] = Workflow::STATUS_DRAFT;
+    $prefix = 'Copy of ';
+    $newName = $prefix . $workflow->getName();
+    $nameColumnLengthInfo = $this->wpdb->get_col_length($this->workflowTable, 'name');
+    $maxLength = is_array($nameColumnLengthInfo)
+      ? $nameColumnLengthInfo['length'] ?? 255
+      : 255;
+    if (strlen($newName) > $maxLength) {
+      $truncateWith = '…';
+      $truncationLength = strlen($truncateWith);
+      $newName = substr($newName, 0, $maxLength - $truncationLength) . $truncateWith;
+    }
+    $data['name'] = $newName;
+    $duplicate = Workflow::fromArray($data);
+    return $this->createWorkflow($duplicate);
+  }
```

Comment:
I'm wondering if duplicating a workflow should also change step IDs. It's not necessary (we never said they are universally unique) but it might be clearer and better for debugging, maybe 🤔

Reply:
One more TODO: the workflow author should now probably be the person who duplicated it.